salt/modules/nagios_json.py

```diff
@@ -0,0 +1,199 @@
+'''
+Check Host & Service status from Nagios via JSON.
+
+.. versionadded:: Beryllium
+
+'''
+
+# Import python libs
+from __future__ import absolute_import
+import logging
+
+# Import 3rd-party libs
+# pylint: disable=import-error,no-name-in-module,redefined-builtin
+from salt.ext.six.moves.urllib.parse import urljoin as _urljoin
+# pylint: enable=import-error,no-name-in-module
+
+try:
+    import requests
+    from requests.exceptions import ConnectionError
+    HAS_DEPS = True
+except ImportError:
+    HAS_DEPS = False
+
+log = logging.getLogger(__name__)
+
+
+def __virtual__():
+    '''
+    Only load if requests is successfully imported
+    '''
+    if not HAS_DEPS:
+        return False
+    return 'nagios_json'
+
+
+def _config():
+    '''
+    Get configuration items for URL, Username and Password
+    '''
+    nagios_url = __salt__['config.get']('nagios_json:nagios_url', '')
+    nagios_username = __salt__['config.get']('nagios_json:nagios_username', '')
+    nagios_password = __salt__['config.get']('nagios_json:nagios_password', '')
+
+    config = {
+        'nagios_url': nagios_url,
+        'nagios_username': nagios_username,
+        'nagios_password': nagios_password,
+    }
+    return config
+
+
+def _status_query(query, method='GET', **kwargs):
+    '''
+    Send query along to Nagios
+    '''
+    headers = {}
+    parameters = {}
+    data = {}
+
+    nagios_url = kwargs.get('nagios_url')
+    nagios_username = kwargs.get('nagios_username')
+    nagios_password = kwargs.get('nagios_password')
+
+    query_params = {
+        'service': [
+            'hostname',
+            'servicedescription',
+        ],
+        'host': [
+            'hostname',
+        ],
+    }
+    parameters['query'] = query
+    for param in query_params[query]:
+        parameters[param] = kwargs[param]
+
+    if not nagios_url.endswith('/'):
+        nagios_url = nagios_url + '/'
+
+    if 'cgi-bin' in nagios_url:
+        url = _urljoin(nagios_url, 'statusjson.cgi')
+    else:
+        url = _urljoin(nagios_url, 'cgi-bin/statusjson.cgi')
+
+    try:
+        if nagios_username and nagios_password:
+            result = requests.request(
+                method=method,
+                url=url,
+                headers=headers,
+                params=parameters,
+                data=data,
+                verify=True,
+                auth=(nagios_username, nagios_password)
+            )
+        else:
+            result = requests.request(
+                method=method,
+                url=url,
+                headers=headers,
+                params=parameters,
+                data=data,
+                verify=True,
+            )
+        if result.status_code == 200:
+            data = result.json()
+        elif result.status_code == 401:
+            log.info('Authentication failed. Check nagios_username and nagios_password.')
+        elif result.status_code == 404:
+            log.info('Url {0} not found.'.format(url))
```

Comment:
I disagree on magic numbers. These codes are actually constants from the `httplib`. Please use them instead of hard-coding 200, 401, 404, even if everyone knows what it is. Instead, please replace them with `httplib.OK`, `httplib.UNAUTHORIZED`, `httplib.NOT_FOUND`  respectively.